stan/math/rev/fun/dot_self.hpp

```diff
@@ -45,10 +20,18 @@ class dot_self_vari : public vari {
  * @param[in] v Vector.
  * @return Dot product of the vector with itself.
  */
-template <typename T, require_eigen_vector_vt<is_var, T>* = nullptr>
+template <typename T, require_rev_matrix_t<T>* = nullptr>
 inline var dot_self(const T& v) {
-  const Eigen::Ref<const plain_type_t<T>>& v_ref = v;
-  return {new internal::dot_self_vari(v_ref)};
+  arena_t<plain_type_t<decltype(value_of(v))>> v_val = value_of(v);
```

Comment:
```suggestion
  arena_t<decltype(value_of(v))> v_val = value_of(v);
```
No need to use `plain_type_t` in `arena_t`. `arena_t<plain_type_t<T>> = arena_t<T>`